File: ProblemSolve/1160.cs
/***********************************************
 * Find Words That Can Be Formed by Characters *
 ***********************************************/

public class Solution {
    public int CountCharacters(string[] words, string chars) {
        int[] countChars = new int[26];             //주어진 chars의 char마다 개수 세기, 일종의 hash
        bool canBeFormed = true;                    //만들 수 있는지
        int ans = 0;
        
        //개수 세기
        foreach(char ch in chars){
            countChars[ch - 97]++;
        }
        
        foreach(string word in words){
            int[] countWordChars = new int[26];
            //각 단어마다 char의 개수를 세서 주어진 chars로 만들 수 있는지 확인
            //chars의 주어진 개수보다 많이 써야하는 경우 false
            foreach(char ch in word){
                countWordChars[ch - 97]++;
                
                if(countChars[ch-97] < countWordChars[ch-97]){
                    canBeFormed = false;
                    break;
                }
            }
            
            if(canBeFormed){                
                ans += word.Length;
            }
            
            canBeFormed = true;
        }
        
        return ans;
    }
}

File: ProblemSolve/1160.cpp
/***********************************************
 * Find Words That Can Be Formed by Characters *
 ***********************************************/

class Solution {
public:
    //주어진 chars에서 각 char의 개수를 세고
    //words의 word마다 char의 개수를 세서 주어진 조건에 부합하는지 확인 후 정답에 추가
    int countCharacters(vector<string>& words, string chars) {
        vector<int> countChars(26);
        bool canBeFormed = true;
        int ans = 0;
        
        //주어진 chars에서 각 char의 개수를 센다
        for(char ch : chars){
            countChars[ch-'a']++;
        }
        
        //words의 word마다 
        for(string word : words){
            vector<int> countWordChars(26);
            
            //char의 개수를 세서 
            for(char ch : word){
                countWordChars[ch-'a']++;
                
                //주어진 조건에 부합하는지 확인 후 
                if(countChars[ch-'a'] < countWordChars[ch-'a']){
                    canBeFormed = false;
                    break;
                }
            }
            
            //정답에 추가
            if(canBeFormed){
                ans += word.size();
            }
            canBeFormed = true;
        }
        
        return ans;
    }
};